src/object.d

```diff
@@ -44,6 +36,119 @@ alias string  = immutable(char)[];
+        return false;
+
+    // This is function is used as a compiler intrinsic and explicitly written
+    // in a lowered flavor to use as few CTFE instructions as possible.
+    size_t idx = 0;
+    immutable length = a1.length;
+
+    for(;idx < length;++idx)
+    {
+        if (a1[idx] != a2[idx])
+            return false;
+    }
+    return true;
+}
+
+/**
+Destroys the given object and sets it back to its initial state. It's used to
+_destroy an object, calling its destructor or finalizer so it no longer
+references any other objects. It does $(I not) initiate a GC cycle or free
+any GC memory.
+*/
+void destroy(T)(ref T obj) if (is(T == struct))
+{
+    // We need to re-initialize `obj`.  Previously, the code
+    // `auto init = cast(ubyte[])typeid(T).initializer()` was used, but
+    // `typeid` is a runtime call and requires the `TypeInfo` object which is
+    // not usable when compiling with -betterC.  If we do `obj = T.init` then we
+    // end up needlessly calling postblits and destructors.  So, we create a
+    // static immutable lvalue that can be re-used with subsequent calls to `destroy`
+    shared static immutable T init = T.init;
+
+    _destructRecurse(obj);
+    () @trusted {
+        auto dest = (cast(ubyte*) &obj)[0 .. T.sizeof];
+        auto src = (cast(ubyte*) &init)[0 .. T.sizeof];
+        dest[] = src[];
+    } ();
+}
+
+private void _destructRecurse(S)(ref S s)
+    if (is(S == struct))
+{
+    static if (__traits(hasMember, S, "__xdtor") &&
+            // Bugzilla 14746: Check that it's the exact member of S.
+            __traits(isSame, S, __traits(parent, s.__xdtor)))
+        s.__xdtor();
+}
+
+nothrow @safe @nogc unittest
+{
+    {
+        struct A { string s = "A";  }
+        A a;
+        a.s = "asd";
+        destroy(a);
+        assert(a.s == "A");
+    }
+    {
+        static int destroyed = 0;
+        struct C
+        {
+            string s = "C";
+            ~this() nothrow @safe @nogc
+            {
+                destroyed ++;
+            }
+        }
+
+        struct B
+        {
+            C c;
+            string s = "B";
+            ~this() nothrow @safe @nogc
+            {
+                destroyed ++;
+            }
+        }
+        B a;
+        a.s = "asd";
+        a.c.s = "jkl";
+        destroy(a);
+        assert(destroyed == 2);
+        assert(a.s == "B");
+        assert(a.c.s == "C" );
+    }
+}
+
+version(D_BetterC) { }
+else
+{
+    version = Not_BetterC;
+}
+
+/*============================ NOT -betterC ====================================
+  Anything below this line should not be imported if compiling with -betterC
+*/
+version(Not_BetterC):
+
+private
+{
+    extern (C) Object _d_newclass(const TypeInfo_Class ci);
+    extern (C) void rt_finalize(void *data, bool det=true);
+}
+
+public @trusted @nogc nothrow pure extern (C) void _d_delThrowable(scope Throwable);
+
 version (D_ObjectiveC) public import core.attribute : selector;
```

Comment:
Using Objective-C together with betterC should be possible.

Reply:
Done. Thanks!